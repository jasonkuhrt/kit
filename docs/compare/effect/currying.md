# Currying Approaches: Kit vs Effect

This document compares how Kit and Effect handle currying and dual APIs for functional programming patterns.

## Overview

Both Kit and Effect need to support two calling styles:

- **Data-first**: `map(array, fn)` - Traditional, familiar style
- **Data-last**: `map(fn)(array)` - Composable, pipe-friendly style

## Kit's Approach: Multiple Named Functions

Kit uses three separate functions with clear naming conventions:

```typescript
// Kit style - 3 separate functions with clear names
map(array, fn) // normal signature
mapOn(array)(fn) // "map on this array" - data first partial
mapWith(fn)(array) // "map with this function" - function first partial

// Usage examples
map([1, 2, 3], x => x * 2) // [2,4,6]
mapOn([1, 2, 3])(x => x * 2) // [2,4,6]
mapWith(x => x * 2)([1, 2, 3]) // [2,4,6]

// In pipes
pipe(
  [1, 2, 3],
  mapWith(x => x * 2),
  filterWith(x => x > 3),
)
```

## Effect's Approach: Single Dual Function

Effect uses a single function that can be called both ways:

```typescript
// Effect style - one function that works both ways
import { dual } from 'effect/Function'

const map = dual(
  2, // arity
  // data-first signature
  <A, B>(self: Array<A>, f: (a: A) => B): Array<B> => {/*...*/},
  // data-last signature
  <A, B>(f: (a: A) => B) => (self: Array<A>): Array<B> => {/*...*/},
)

// Usage - same function, different call patterns
map([1, 2, 3], x => x * 2) // data-first
map(x => x * 2)([1, 2, 3]) // data-last via currying

// In pipes
pipe(
  [1, 2, 3],
  map(x => x * 2), // Same function!
  filter(x => x > 3),
)
```

### Effect's Implementation

Effect uses runtime arity checking to determine which version to use:

```typescript
// Simplified version of Effect's dual
function dual(arity, dataFirst, dataLast) {
  return function(...args) {
    if (args.length >= arity) {
      return dataFirst(...args)
    } else {
      return dataLast(...args)
    }
  }
}
```

## Key Difference: Missing Pattern

**Effect has no equivalent to Kit's `*On` pattern**. Effect's dual functions only support:

- Full application: `map(array, fn)`
- Function-first partial: `map(fn)` (like Kit's `*With`)

But NOT:

- Data-first partial: `map(array)` ❌ (would be Kit's `*On`)

This means Effect cannot express certain patterns that Kit can.

## Comparison

### Kit's Advantages

1. **Explicit intent**: `mapOn` vs `mapWith` clearly shows what's partially applied
2. **Better TypeScript inference**: Each function has one clear signature
3. **No runtime overhead**: No need to check arity at runtime
4. **Easier to understand**: Function name tells you exactly how to use it
5. **Tree-shaking friendly**: If you only use `map`, you don't bundle `mapOn` and `mapWith`
6. **Principle of least surprise**: Each function does exactly one thing
7. **More partial application options**: Can partially apply either argument

### Effect's Advantages

1. **Less API surface**: One function name instead of three
2. **Familiar to FP users**: Similar to Ramda's approach
3. **Consistent naming**: Don't need to remember `On` vs `With` patterns
4. **Migration friendly**: Easier to convert existing code

## Examples in Practice

### Kit Style

```typescript
// For piping - function first (*With pattern)
const doubleAll = mapWith(x => x * 2) // (array) => array
const numbersOnly = filterWith(isNumber) // (array) => array
const firstFive = takeWith(5) // (array) => array

pipe(
  [1, 2, 3, 4, 5, 6],
  doubleAll,
  numbersOnly,
  firstFive,
)

// For partial application with data (*On pattern)
const mapOnUsers = mapOn(users) // (fn) => result
const filterOnPosts = filterOn(posts) // (predicate) => result

// Apply different operations to the same data
const userNames = mapOnUsers(user => user.name)
const userAges = mapOnUsers(user => user.age)
const activePosts = filterOnPosts(post => post.isActive)
const draftPosts = filterOnPosts(post => post.isDraft)
```

### Effect Style

```typescript
// Only supports function-first partial application
const doubleAll = map(x => x * 2)
const numbersOnly = filter(isNumber)
const firstFive = take(5)

// Cannot partially apply data!
const mapOnUsers = map(users) // ❌ This will error - expects a function
// Effect's dual would interpret this as partial application waiting for second arg

// Must use closure instead
const mapOnUsers = (fn) => map(users, fn) // Manual wrapper needed
```

## Conclusion

Kit's approach trades verbosity for clarity. While Effect's dual functions reduce API surface, Kit's explicit naming makes code more readable and intentions clearer. The performance benefit of avoiding runtime arity checks is a nice bonus.

For a utility library meant to be approachable and tree-shakeable, Kit's design choice aligns well with its goals.
