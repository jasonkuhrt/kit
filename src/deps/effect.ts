export { Match as M, Schema as S } from 'effect'
