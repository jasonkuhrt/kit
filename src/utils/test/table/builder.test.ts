import { Test } from '#test'
import { Effect, Layer } from 'effect'
import { describe, expect, it } from 'vitest'

// Test functions for testing
const add = (a: number, b: number): number => a + b
const multiply = (a: number, b: number): number => a * b
const upperCase = (s: string): string => s.toUpperCase()
const identity = <T>(x: T): T => x
const constant = () => 42

describe('Test.table() builder', () => {
  // These tests run actual test cases (outside of it() blocks)
  describe('functional tests', () => {
    // Test default assertion (toEqual)
    Test.describe('default toEqual assertion')
      .on(add)
      .cases(
        [[1, 2], 3],
        [[5, 5], 10],
      )

    // Test structural equality with toEqual
    const merge = (a: any, b: any) => ({ ...a, ...b })
    Test.describe('structural equality')
      .on(merge)
      .cases(
        [[{ a: 1 }, { b: 2 }], { a: 1, b: 2 }], // structural equality works with toEqual
      )

    // Test custom assertion
    Test.describe('custom assertion')
      .on(identity)
      .cases(
        [[42], 42],
      )
      .test((actual, expected) => {
        expect(actual).toBe(expected) // reference equality
      })

    // Test casesAsArg
    Test.describe('casesAsArg helper')
      .on(upperCase)
      .casesAsArgs(
        'hello',
        'world',
      )
      .test()

    // Test multi-argument snapshot with separator
    Test.describe('multi-argument snapshots')
      .on(add)
      .casesAsArgs(
        [1, 2],
        [10, 20],
      )
      .test()
  })

  // These tests verify the API without executing (inside it() blocks)
  describe('basic functionality', () => {
    it('should compile with basic cases', () => {
      // This is a compile-time test - just verify it compiles
      const builder = Test.describe()
        .inputType<string>()
        .outputType<string>()
      // Verify the builder has the expected methods
      expect(builder.cases).toBeDefined()
      // test() only exists after cases()
      const withCases = builder.cases({ n: 'test', i: 'input', o: 'output' })
      expect(withCases.test).toBeDefined()
    })

    it('should work with description', () => {
      // This is a compile-time test - just verify it compiles
      const builder = Test.describe('string transformations')
        .inputType<string>()
        .outputType<string>()
      // Verify the builder has the expected methods
      expect(builder.cases).toBeDefined()
      // test() only exists after cases()
      const withCases = builder.cases({ n: 'test', i: 'input', o: 'output' })
      expect(withCases.test).toBeDefined()
    })
  })

  describe('.on() functionality', () => {
    it('should work with multi-param functions', () => {
      // Verify the API compiles and builder is returned
      const builder = Test.describe()
        .on(add)
      expect(builder.cases).toBeDefined()
      expect(builder.case).toBeDefined()
    })

    it('should work with single-param functions', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .on(upperCase)
      expect(builder.cases).toBeDefined()
      expect(builder.casesAsArgs).toBeDefined()
    })

    it('should work with zero-param functions', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .on(constant)
      expect(builder.cases).toBeDefined()
    })

    it('should support snapshot mode', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .on(upperCase)
      expect(builder.cases).toBeDefined()
    })

    it('should support .case() method', () => {
      // Verify the API compiles and chaining works
      const builder = Test.describe()
        .on(add)
        .case(2, 3, 5) // Direct params + output
        .case('named', 5, 5, 10) // Named with direct params + output
      expect(builder.test).toBeDefined()
    })

    it('should support custom assertions in .test()', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .on(add)
        .cases([[1, 2], 3])
      expect(builder.test).toBeDefined()
    })
  })

  describe('type building', () => {
    it('should work with .i() and .o()', () => {
      // This is a compile-time test - just verify it compiles
      const builder = Test.describe()
        .inputType<string>()
        .outputType<string>()
      expect(builder.cases).toBeDefined()
      // test() only exists after cases()
      const withCases = builder.cases({ n: 'test', i: 'input', o: 'output' })
      expect(withCases.test).toBeDefined()
    })
  })

  describe('modifiers', () => {
    it('should support skip and todo', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .on(add)
      expect(builder.skip).toBeDefined()
      expect(builder.skipIf).toBeDefined()
    })

    it('should support only', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .only()
        .on(add)
      expect(builder.cases).toBeDefined()
    })

    it('should support skipIf', () => {
      const shouldSkip = true
      // Verify the API compiles
      const builder = Test.describe()
        .skipIf(() => shouldSkip)
        .on(add)
      expect(builder.cases).toBeDefined()
    })
  })

  describe('Effect support', () => {
    // This is a compile-time test to ensure the APIs exist
    // Actual Effect testing would require proper setup

    it('should compile with .layer()', () => {
      // This test just ensures the API compiles
      const code = () => {
        const mockLayer = Layer.empty as Layer.Layer<any>
        Test.describe()
          .inputType<number>()
          .outputType<number>()
          .cases(
            { n: 'test', i: 1, o: 2 },
          )
          .layer(mockLayer)
          .testEffect(() => Effect.succeed(undefined))
      }
      expect(code).toBeDefined()
    })

    it('should compile with .layerEach()', () => {
      // This test just ensures the API compiles
      const code = () => {
        Test.describe()
          .inputType<number>()
          .outputType<number>()
          .cases(
            { n: 'test 1', i: 1, o: 2 },
          )
          .layerEach(({ i }) => Layer.empty)
          .testEffect(() => Effect.succeed(undefined))
      }
      expect(code).toBeDefined()
    })
  })

  describe('custom matchers', () => {
    it('should support .onlyMatching()', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .on(identity)
        .onlyMatching('toEqual')
      expect(builder.cases).toBeDefined()
    })
  })

  describe('snapshot testing', () => {
    it('should support snapshot testing', () => {
      // This is a compile-time test - verify snapshot API exists
      const builder = Test.describe()
        .inputType<string>()
      expect(builder.cases).toBeDefined()
      // test() only exists after cases()
      const withCases = builder.cases({ n: 'test', i: 'input', o: undefined })
      expect(withCases.test).toBeDefined()
    })
  })

  describe('tuple vs object cases', () => {
    it('should accept all tuple formats', () => {
      // Verify the API compiles with various tuple formats
      const builder = Test.describe()
        .on(add)
      // Just verify it has the methods
      expect(builder.cases).toBeDefined()
    })

    it('should accept all object formats', () => {
      // Verify the API compiles with various object formats
      const builder = Test.describe()
        .on(add)
      // Just verify it has the methods
      expect(builder.cases).toBeDefined()
    })
  })

  describe('concurrent execution', () => {
    it('should support .concurrent()', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .concurrent()
        .on(add)
      expect(builder.cases).toBeDefined()
    })
  })

  describe('name templates', () => {
    it('should support .name() template', () => {
      // Verify the API compiles
      const builder = Test.describe()
        .on(add)
        .name('$i[0] + $i[1] = $o')
      expect(builder.cases).toBeDefined()
    })
  })

  describe('progressive test building', () => {
    it('should support building tests incrementally', () => {
      const suite = Test.describe()
        .on(multiply)

      // Add cases incrementally
      suite.case(2, 3, 6) // Direct params
      suite.case('double', 5, 2, 10) // Named with direct params
      suite.case(10, 10) // Direct params for snapshot
      // Cases will execute at the end of the describe block
    })
  })
})
