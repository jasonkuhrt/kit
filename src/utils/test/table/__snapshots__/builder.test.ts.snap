// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test.table() builder > functional tests > casesAsArg helper > upperCase("hello") 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
hello
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
HELLO
╚══════════════════════════════════════════════════╝"
`;

exports[`Test.table() builder > functional tests > casesAsArg helper > upperCase("world") 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
world
╠══════════════════════════════════════════════════╣ THEN RETURNS STRING
WORLD
╚══════════════════════════════════════════════════╝"
`;

exports[`Test.table() builder > functional tests > multi-argument snapshots > add(1, 2) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
1
────────────────────────────────────────────────────
2
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
3
╚══════════════════════════════════════════════════╝"
`;

exports[`Test.table() builder > functional tests > multi-argument snapshots > add(10, 20) 1`] = `
"
╔══════════════════════════════════════════════════╗ GIVEN ARGUMENTS
10
────────────────────────────────────────────────────
20
╠══════════════════════════════════════════════════╣ THEN RETURNS NUMBER
30
╚══════════════════════════════════════════════════╝"
`;
